### PR TITLE
A first draft of Lattice related objects

### PR DESCRIPTION
**Pull request by bollu**
1. I don't understand the split between `src/Data` and `src/Order`. Can you shed some light on what goes where?
2. Related to the above: where do I throw in CPOs and DCPOs?
3. There are different representations for an "arbitrary finite partial order" with different trade-offs. 
       - One could store a DAG, paying `O(n)` space but also `O(non-constant)` time for queries.
       - One could store the matrix of all possible comparisons, paying `O(n^2)` space but `O(1)` time for queries. 
    Is implementing a general purpose representation of arbitrary finite partial orders/lattice a goal of this library?
4. With respect to Heytig algebras, How do we want to motivate the axioms? I don't have a great story

### CODE REVIEW

**Comment by emilypi** [comment]
1. Originally `Data` was `Algebra` and I wasn't sure how much lattice stuff i'd have to magpie from other libraries. It's significant enough now that we should probably just hurl everything in Data, and reorg `Order.Partial` to `Data.Ord.(Partial|Pre)` so we keep the namespace consistent. 
2. In a new `Data.Ord.Partial` namespace (e.g. `Data.Ord.Partial.Complete` choose your fighter here. Could be CPO instead of Complete). 
3. Let's think on this one
4. We could do this in myriad ways: 

  - Logically: A heyting algebra is a deductive system that admits a Truthiness, Falsity, conjunctions, dysjunctions, and modus 
    ponens arguments (a, a -> b |- b). In other words - the axioms of intuitionistic logic hold. 
  - Algebraically: Heyting algebras `(L, 0, 1, ^, v, ->)`  are bounded lattices equipped with a binary operation called implication 
    (->), for which every residue `a -> (-) ≝ a ^(-)` exists. 
  - Categorically: Heyting algebras are Bounded Lattices (seen as a category) that admit currying (exponentiation has a left adjoint). 

  As we're writing the lattice/order stuff, i'd like to be aware of how we're accumulating structure, with the end result being _eventually_ an intuitionistic system or CCC.

**Comment by bollu** [comment]
- For Heytig algebras, I went with the "has an implication", since that's the most (computationally) interesting version.
- Now I'm stuck on CPOs. I don't know of a computationally relevant definition, or a useful example to litmus test any definition I choose. Do you have any ideas?

File: src/Data/Order/Complete.hs [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+-- | Davey and Priestly, 2nd edition, Definition 7.7: Directed set:
+-- let S be a non-empty subset of an ordered set P.
+-- Then ∀ x, y ∈ S, ∃z ∈ S, z ∈ x ≤ z ∧ y ≤ z
+
+-- | Davey and Priestly, 2nd edition, Definition 8.1: a CPO is an ordered set with a bottom, 
+-- where each directed set has a supremum.
+class PartialOrd a => CompletePartialOrd a where
+   -- | conceivably, some types may have better encodings of directed subsets
+   -- than others..?
+   type Directed a -- associated type of directed subsets?
```
Comment (emilypi) [inline src/Data/Order/Complete.hs:14]: makes sense

**Review by emilypi — APPROVED** [review]
Good for now, sans comments

**Pull request merged**